smaug/python/ops/control_flow_ops.py

```diff
@@ -30,9 +28,100 @@ def merge(input_tensors, name="merge"):
+    """Construct a control flow context.
+
+    Args:
+      pred: Predication tensor.
+      branch: Value of 0 or 1 to denote the False/True branch of the control
+        flow.
+    """
+    self._pred = pred
+    self._branch = branch
+    # Names of all nodes that are within this context.
+    self._internal_nodes = set()
+    # Mappings of the name of a known external tensors to its switch node ouput.
+    self._external_tensors = {}
+    # The next outer context to this context.
+    self._outer_context = None
+
+  def __enter__(self):
+    """Enter this control flow context."""
+    self._outer_context = global_vars.get_graph().control_flow_context
+    global_vars.get_graph().control_flow_context = self
+    return self
+
+  def __exit__(self, *args):
+    """Exit this control flow context."""
+    global_vars.get_graph().control_flow_context = self._outer_context
+
+  def add_result(self, result):
+    """Make the result nodes available in the outer context.
+
+    The results of the inner context will be used in the outer context. To
+    prevent the outer context from thinking that these results are external
+    and consequently adding unnecessary switch nodes, this makes the inner
+    nodes that produce the results internal to outer context.
+    """
+    if self._outer_context:
+      for res in result:
+        self._outer_context.add_node_name(res.source[0].name)
+
+  def add_node_name(self, name):
+    self._internal_nodes.add(name)
+
+  def add_node(self, node, input_tensors):
+    """Add this node to the control flow context.
+
+    This makes the node known to the context. We create a switch node for any
+    of its input tensor if it's deemed external to the context.
+    """
+    # Don't add the node if it's already added. Don't add a data op to avoid
+    # deadlock.
+    if node.name in self._internal_nodes or node.op == types_pb2.Data:
+      return input_tensors
+    ret_tensors = []
+    for tensor in input_tensors:
+      # If this tensor comes from an internal node or if we already added a
+      # switch node for this tensor, don't create a switch node.
+      if tensor.source != None and (tensor.source[0].name
+                                    in self._internal_nodes):
+        ret_tensors.append(tensor)
+      elif tensor.name in self._external_tensors:
+        ret_tensors.append(self._external_tensors[tensor.name])
+      else:
+        switch_name = global_vars.get_graph().create_unique_name(
+            "switch", mark_as_used=False)
+        # To avoid deadlock in creating switch nodes, make the switch node
+        # internal to the context before it's created.
+        self._internal_nodes.add(switch_name)
+        ret_tensor = switch(tensor, self._pred, name=switch_name)[self._branch]
+        ret_tensors.append(ret_tensor)
+        self._external_tensors[tensor.name] = ret_tensor
+    self._internal_nodes.add(node.name)
+    return ret_tensors
+
+def cond(pred, true_fn, false_fn, name="cond"):
```

Comment:
I'd prefer "predicate" or "condition" over "pred", which is often used in other circumstances (like predecessor).

Also, the fact that you use lambdas as arguments to cond() is very unintuitive and inconsistent with the rest of the API. Why do this? At first, I thought you wanted to avoid creating the ops if that branch isn't taken, but both ops have to be constructed anyways since SMAUG uses deferred execution. Then I thought maybe you want to hide the op inside the ControlFlowContext, but I don't understand what benefits that provides. 

Overall, the use of lambdas in the cond() interface is highly unusual and I would not support it without really good reasons.

Reply:
Changed `pred` to `predication`.